r/R/query-engine.R

```diff
@@ -101,7 +101,11 @@ ExecPlan <- R6Class("ExecPlan",
         # plus group_by_vars (last)
         # TODO: validate that none of names(aggregations) are the same as names(group_by_vars)
         # dplyr does not error on this but the result it gives isn't great
-        node <- node$Project(summarize_projection(.data))
+        projection <- summarize_projection(.data)
+        # skip projection if no grouping and all aggregate functions are nullary
```

Comment:
Why?

Reply:
In general we want to avoid adding unnecessary projections to the ExecPlan. This is a case where we can do that.

Reply:
Would something be harmed if it were projected to having 0 columns? Would that be even more efficient?

Reply:
My understanding is that adding a project node before an aggregate node will never improve performance, and that the only real reason to do it is to ensure that the aggregate node has all the columns it needs.

I just checked with Weston about this. Here's our conversation:

**Ian:**
Which of these two ExecPlans would execute the most quickly and efficiently?

1. Read source data containing columns A,B,C,D --> project to only columns A and B --> aggregate sum of A over groups in B --> output to table

2. Read source data containing columns A,B,C,D --> aggregate sum of A over groups in B --> output to table

In other words, does adding an unnecessary projection help, hurt, or not make any difference?

**Weston**:
I assume by "project to" you mean an actual project node and not a push down projection? In reality you'd want to avoid reading C and D from disk so you'd want the projection in the read.

**Ian**:
correct, actual project node

**Weston**:
An actual project node, with just field references, is effectively free.

Now, if the follow-up node was a filter node, then you might be paying more for plan 2.  We don't use selection vectors so the filter node would have to materialize the data and it would have to materialize your two junk columns.

The aggregate node, on the other hand, doesn't output any input columns.  So I'd expect it will throw them away in it's InputReceived and the performance of 1 & 2 will be more or less the same.